### PR TITLE
Add promisified methods to the adapter class

### PR DESCRIPTION
**Pull request by AlCalzone**
As we only support NodeJS 4+ with native promise support, we can now natively provide promisified versions of the `Adapter` class methods. These have the name of the original function with `Async` appended to it. Adapter developers can now opt to use promises instead of classic callbacks - or even `async`/`await` when using TypeScript, e.g.:
```js
adapter.getStateAsync("foo").then(result => {
    // do something with the result
}).catch(e => {
    // something went wrong
});

// or even this:
let fooValue;
try {
    fooValue = await adapter.getStateAsync("foo");
    // do something with the result
} catch (e) { 
    // something went wrong
}
```

I exposed a bunch of adapter methods (including the most important `setState`, `getState`, etc...). Other ones (especially the ones linked to objects.js) are still missing.

**Before I invest more time, what are your thoughts on this?**

Also, according to http://node.green/#ES2015-functions-class we should be able to rewrite the Adapter class using the `class` keyword, since this is supported in NodeJS 4 aswell. It would greatly improve the auto-completion and hints when using VSCode for example. I can give it a shot aswell if you're okay with that.

### CODE REVIEW

**Comment by AlCalzone** [comment]
So I just replaced my JS-Controller with the one from this PR and also changed ioBroker.tradfri to a branch where my locally promisified versions of `{get,set}{State,Object}` were replaced with the new version from js-controller/lib/adapter.js.
Works like a charm!

**Comment by GermanBluefox** [comment]
I wanted to plan this for a long time and to implement promises in JS-Adapter too. But still no time.
It is perfect, that you found the time for that.
We should add according tests for Promises.

**Comment by AlCalzone** [comment]
I think I got all the methods in adapter.js now (except for some weird ones I don't understand like `pushFiFo` and similar ones). 

How would you build the according tests? Or is that something for our test master @Apollon77 ?

**Comment by Apollon77** [comment]
Hehe ;-)  if I would understand and use promises I maybe would do it ;-)

Basically I would say: copy the existing tests and rewrite to promise style. All files that Start with „test“ will be executed by the tests ...

**Comment by Apollon77** [comment]
Ps: please check Travis-ci because currently tests are failing for nodejs 4 in your pr branch

**Comment by Apollon77** [comment]
Ignore last comment ... went green again ;-))

**Comment by AlCalzone** [comment]
So I played around a bit with the async method tests, but this requires more work.
The tests which interact with the objects DB cannot simply be copied after the sync tests because after the sync tests some variables exist which did not before. As a result some assertions in the async tests fail.

Testing the same functionality sync and async will require some rewriting of the test suite. Or maybe run the same again with only async functions - dunno?!

**Comment by Apollon77** [comment]
Ok, this should not be the problem at all ... we could do it comparable to the notmal adapter test. There the objects/states.json is copied before the test starts and can be restored to it's initial value ...

like: https://github.com/Apollon77/ioBroker.smartmeter/blob/master/test/lib/setup.js#L81 ff ?!

**Comment by AlCalzone** [comment]
As you're the test master, could you make me a framework where I can fill in the actual tests? Just create a PR on my branch ("promises").

**Comment by Apollon77** [comment]
I can try, but needs time! sick currently ... only small things possible

File: lib/adapter.js [hunk 1]
```diff
@@ -22,11 +22,11 @@ var fs =                require('fs');
 var extend =            require('node.extend');
 var util =              require('util');
 var EventEmitter =      require('events').EventEmitter;
-var tools =             require('./tools');
+var tools =             require(__dirname + '/tools');
```
Comment (AlCalzone) [inline lib/adapter.js]: What's the reason you're using `__dirname` over relative paths (`'./abc'`)?

**Pull request merged**